doc/arch/adr-002-geoprocessing-caching.md

```diff
@@ -0,0 +1,103 @@
+
+While we could cache the entire output of a MapShed or TR-55 run, in the case of an update to some of the constituent raster or vector data would force recalculation of the whole. By caching only the time-taking geoprocessing results, we ensure that any updates to constituent rasters would invalidate only that specific cached result, leaving the others still current. And since vector data results would never be cached, they will always be current upon update as well.
+
+In case we do not foresee updating the rasters very often, it may be beneficial to consider caching the entire output.
+
+### Where will it be cached?
+
+The current app stack gives us two options to choose as a cache: Redis and PostgreSQL. While reads from Redis can potentially be a bit faster, the in-memory cache will be purged upon restart and deployment. Furthermore, the use case to optimize is a faster read by one user over time, rather than parallel reads by many users concurrently, the speed advantages of Redis are of little use.
+
+By storing the cached results in the database, we ensure they are valid and available after deployments and new releases, while still being considerably faster than before.
+
+### When will it be cached?
+
+When a user selects a WKAoI, and a request is made to `/analyze` or `/modeling`, if the WKAoI results haven't already been cached, we will run and cache them. Over time, we will build up the cache, so that when new users request the same WKAoIs, they will get the cached results.
+
+For a select few WKAoIs which are too large to process in the production infrastructure, such as HUC-8s which time-out during MapShed gathering phase, we can run their geoprocessing steps on more powerful infrastructure with longer timeouts in a batch process, cache the results, and then decommission it. This will make them available to regular users of the production app without needing the extra power to render them.
+
+### How will the items be identified, cached, and retrieved?
+
+Each geoprocessing run can be identified by two things: 1) input shape, and 2) JSON package sent to geoprocessing service. Because the input shape is a WKAoI, we can use it's table name and id as the identifier, for example `boundary_huc08__1`. The JSON package can be identified by the key in [`settings.GEOP['json']`](https://github.com/WikiWatershed/model-my-watershed/blob/a49cc115d18d2c67cbd0721cf62faa08e98d13fc/src/mmw/mmw/settings/base.py#L387), such as `nlcd_streams`, `nlcd_soils`, `slope`, etc.
+
+We can have a table structure as follows:
+
+| Column     | Type                     | Modifiers |
```

Comment:
Yeah. I'm also not sure how we would do out-of-band caching of very large WKAoIs manually. Let me see if these operations can be supported by the low level cache API, but otherwise we may end up going with a custom table.